h/claim/views.py

```diff
@@ -0,0 +1,115 @@
+import logging
+
+import deform
+from horus.lib import FlashMessage
+from pyramid import httpexceptions as exc
+from pyramid.view import view_config
+
+from . import schemas
+from ..models import _
+from ..accounts.models import User
+from ..accounts.events import LoginEvent
+
+
+log = logging.getLogger(__name__)
+
+
+@view_config(route_name='claim_account',
+             request_method='GET',
+             renderer='h:claim/templates/claim_account.html')
+def claim_account(request):
+    _validate_request(request)
+
+    form = _form_for_update_account(request)
+    return {'form': form}
+
+
+@view_config(route_name='claim_account',
+             request_method='POST',
+             renderer='h:claim/templates/claim_account.html')
+def update_account(request):
+    user = _validate_request(request)
+
+    form = _form_for_update_account(request)
+    try:
+        appstruct = form.validate(request.POST.items())
+    except deform.ValidationFailure:
+        return {'form': form}
+
+    # The token is valid and the form validates, so we can go ahead and claim
+    # the account:
+    user.password = appstruct['password']
+
+    msg = _('Your account has been successfully claimed.')
+    FlashMessage(request, msg, kind='success')
+
+    request.registry.notify(LoginEvent(request, user))
+    return exc.HTTPFound(location=request.route_url('stream'))
+
+
+def _validate_request(request):
+    """
+    Check that the passed request is appropriate for proceeding with account
+    claim. Asserts that:
+
+    - no-one is logged in
```

Comment:
We automatically log them in at the end of the claim process, so it seems sensible to require that they start from a logged-out state so that they're not confused about who they're logged in as at the end.